crates/sui-core/src/authority/backpressure.rs

```diff
@@ -0,0 +1,314 @@
+        Arc::new(Self {
+            watermarks_sender,
+            backpressure_sender,
+        })
+    }
+
+    pub fn new_from_checkpoint_store(store: &CheckpointStore) -> Arc<Self> {
+        let executed = store
+            .get_highest_executed_checkpoint_seq_number()
+            .expect("read cannot fail")
+            .unwrap_or_default();
+        let certified = store
+            .get_highest_synced_checkpoint_seq_number()
+            .expect("read cannot fail")
+            .unwrap_or_default();
+        info!(
+            ?executed,
+            ?certified,
+            "initializing backpressure manager from checkpoint store"
+        );
+        Self::new_from_watermarks(Watermarks {
+            executed,
+            certified,
+        })
+    }
+
+    pub fn update_highest_certified_checkpoint(&self, seq: CheckpointSequenceNumber) {
+        self.watermarks_sender.send_if_modified(|watermarks| {
+            if seq > watermarks.certified {
+                debug_assert_eq!(seq, watermarks.certified + 1);
+                watermarks.certified = seq;
+                debug!(?watermarks, "updating highest certified checkpoint");
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    pub fn update_highest_executed_checkpoint(&self, seq: CheckpointSequenceNumber) {
+        self.watermarks_sender.send_if_modified(|watermarks| {
+            if seq > watermarks.executed {
+                debug_assert_eq!(seq, watermarks.executed + 1);
+                watermarks.executed = seq;
+                debug!(?watermarks, "updating highest executed checkpoint");
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    pub fn set_backpressure(&self, backpressure: bool) {
+        self.backpressure_sender.send_if_modified(|bp| {
+            if *bp != backpressure {
+                debug!(?backpressure, "setting backpressure");
+                *bp = backpressure;
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    pub fn subscribe(self: &Arc<Self>) -> BackpressureSubscriber {
+        BackpressureSubscriber { mgr: self.clone() }
+    }
+}
+
+impl BackpressureSubscriber {
+    pub fn is_backpressure_active(&self) -> bool {
+        *self.mgr.backpressure_sender.borrow()
+    }
+
+    pub async fn await_backpressure(&self) {
+        let mut watermarks_rx = self.mgr.watermarks_sender.subscribe();
+        if watermarks_rx
+            .borrow_and_update()
+            .is_backpressure_suppressed()
+        {
+            return;
+        }
+
+        let mut backpressure_rx = self.mgr.backpressure_sender.subscribe();
+        if !*backpressure_rx.borrow_and_update() {
+            return;
+        }
+
+        info!("waiting for backpressure to be lifted");
+        let _scope = monitored_scope("await_backpressure");
+
+        loop {
```

Comment:
I think it will be useful to have a counter on how many times this is blocking, or something like that.

Reply:
That's what the monitored scope is for. Or did you mean that you think there should be a counter for iterations of the loop here? I think its probably not necessary - we can detect that a validator is stuck at this point with the monitored scope, and then enable debug logging to determine whether the loop is advancing or not.